### PR TITLE
delay mempool clearing

### PR DESCRIPTION
**Pull request by rgarc**
https://github.com/monad-crypto/monad-internal/issues/142

This PR aims to decouple the clearing of the mempool from the proposal creation path by introducing a new event and command pair, enabling the mempool to be cleared via the loopback executor.

### CODE REVIEW

File: monad-consensus-state/src/command.rs [hunk 1]
```diff
@@ -56,12 +55,7 @@ where
     StateRootHash(Block<SCT>),
     // TODO-2 add command for updating validator_set/round
     // - to handle this command, we need to call message_state.set_round()
-    //
-    /// Sends transactions remaining in TxPool to another validator
-    CascadeTxns {
-        peer: NodeId<SCT::NodeIdPubKey>,
-        txns: Bytes,
-    },
+    ClearMempool,
```
Comment (michael-yxchen) [inline monad-consensus-state/src/command.rs:61]: nit: update comment

File: monad-executor-glue/src/lib.rs [hunk 2]
```diff
@@ -357,6 +344,7 @@ where
             }
             MonadEvent::AsyncStateVerifyEvent(_) => "ASYNCSTATEVERIFY".to_string(),
             MonadEvent::MetricsEvent(_) => "METRICS".to_string(),
+            MonadEvent::MempoolEvent(MempoolEvent::Clear) => "CLEARMEMPOOL".to_string(),
```
Comment (michael-yxchen) [inline monad-executor-glue/src/lib.rs]: nit: group it with the other mempool event

File: monad-proto/proto/message.proto [hunk 3]
```diff
@@ -61,7 +57,6 @@ message ProtoMonadMessage {
     ProtoUnverifiedConsensusMessage consensus = 1;
     ProtoRequestBlockSyncMessage block_sync_request = 2;
     ProtoBlockSyncMessage block_sync_response = 3;
-    ProtoCascadeTxMessage cascade_txns = 4;
     ProtoPeerStateRootMessage peer_state_root = 5;
```
Comment (michael-yxchen) [inline monad-proto/proto/message.proto]: update `peer_state_root` field number to 4

**Comment by rgarc** [comment]
> while the txpool clear event loopsback, new txns could have been inserted by rpc and could get into txpool first and then get cleared. one approach, txpool could have an active and secondary set of txns and at the end of proposal creation, swap active-secondary and then the clear command only clears secondary.

Thanks for bringing that up, didn't really think of that. Implemented what I think you meant in e0674a1058e6142b239f1c4c65fccf8ed971ada3 but now I'm wondering if, at swap time, secondary can already contain transactions (in other words, the assert below failing). I don't think that would be desirable.
https://github.com/monad-crypto/monad-bft/blob/e0674a1058e6142b239f1c4c65fccf8ed971ada3/monad-eth-txpool/src/lib.rs#L68

**Comment by rgarc** [comment]
> > while the txpool clear event loopsback, new txns could have been inserted by rpc and could get into txpool first and then get cleared. one approach, txpool could have an active and secondary set of txns and at the end of proposal creation, swap active-secondary and then the clear command only clears secondary.
> 
> Thanks for bringing that up, didn't really think of that. Implemented what I think you meant in [e0674a1](https://github.com/monad-crypto/monad-bft/commit/e0674a1058e6142b239f1c4c65fccf8ed971ada3) but now I'm wondering if, at swap time, secondary can already contain transactions (in other words, the assert below failing). I don't think that would be desirable.
> 
> https://github.com/monad-crypto/monad-bft/blob/e0674a1058e6142b239f1c4c65fccf8ed971ada3/monad-eth-txpool/src/lib.rs#L68

> > while the txpool clear event loopsback, new txns could have been inserted by rpc and could get into txpool first and then get cleared. one approach, txpool could have an active and secondary set of txns and at the end of proposal creation, swap active-secondary and then the clear command only clears secondary.
> 
> Thanks for bringing that up, didn't really think of that. Implemented what I think you meant in [e0674a1](https://github.com/monad-crypto/monad-bft/commit/e0674a1058e6142b239f1c4c65fccf8ed971ada3) but now I'm wondering if, at swap time, secondary can already contain transactions (in other words, the assert below failing). I don't think that would be desirable.
> 
> https://github.com/monad-crypto/monad-bft/blob/e0674a1058e6142b239f1c4c65fccf8ed971ada3/monad-eth-txpool/src/lib.rs#L68

Did some more thinking. ~Another downside to this active/secondary approach is that it's non-trivial to make generic over the txpool data structure(s), which is very likely to change in the future.~

Might be easier if we defer destruction to a separate thread. But I wasn't sure if that was reasonable 

something like
```rust
struct TxPool(/* ... */);

fn create_proposal(&mut self) {
    // create_proposal logic

    let mut old_pool = std::mem::take(&mut self.0);
    std::thread::spawn(move || {
        drop(old_pool);
    });

    FullTransactionList::empty()
}
```
Would need to find something such that thread spawning is cheap. This would also obviate the need for mempool clearing events/commands.

**Comment by rgarc** [comment]
### Change Since Previous Benchmark

image

### Understanding this report:

The plot on the left shows the probability of the function taking a certain amount of time. The red curve represents the saved measurements from the last time this benchmark was run, while the blue curve shows the measurements from this run. The lines represent the mean time per iteration. Click on the plot for a larger view.

**Pull request merged**